js/ai/src/resource.ts

```diff
@@ -121,7 +126,25 @@ export function defineResource(
           message: `input ${input} did not match template ${uri}`,
         });
       }
-      return await fn(templateMatch, ctx);
+      const parts = await fn(input, ctx);
+      parts.map((p) => {
+        if (!p.metadata) {
+          p.metadata = {};
+        }
+        if (p.metadata?.resource) {
+          (p.metadata as any).resource.parent = {
+            uri: input,
+            ...(p.metadata as any).resource.parent,
+          };
+        } else {
```

Comment:
How about:

1. if metadata.resource.parent already exists, leave it alone entirely
2. if metadata.resource already exists, populate metadata.resource.parent with {uri,template}
3. if metadata.resource doesn't exist, populate metadata.resource with {uri,template}

Reply:
Done.